### PR TITLE
Stdlib Json

### PR DESCRIPTION
**Pull request by stefan-aws**
### How has this been tested?

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyStandardLibraries/src/dfyconfig.toml [hunk 1]
```diff
@@ -8,10 +8,24 @@ excludes = ["**/Arithmetic/*.dfy", "**/Arithmetic/**/*Internals.dfy"]
...
+  "DafnyStdLibs/Arithmetic/Mul.dfy",
+  "DafnyStdLibs/Arithmetic/Power.dfy",
+  "DafnyStdLibs/Arithmetic/Power2.dfy",
+  "DafnyStdLibs/Arithmetic/Internals/DivInternals.dfy",
+  "DafnyStdLibs/Arithmetic/Internals/DivInternalsNonlinear.dfy",
+  "DafnyStdLibs/Arithmetic/Internals/GeneralInternals.dfy",
+  "DafnyStdLibs/Arithmetic/Internals/ModInternals.dfy",
+  "DafnyStdLibs/Arithmetic/Internals/ModInternalsNonlinear.dfy",
+  "DafnyStdLibs/Arithmetic/Internals/MulInternals.dfy",
+  "DafnyStdLibs/Arithmetic/Internals/MulInternalsNonlinear.dfy",
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/src/dfyconfig.toml:23]:
```suggestion
  "build/DafnyStandardLibrary-arithmetic.doo"
```

Didn't think of this when I originally suggested `library` :)

File: Source/DafnyCore/Rewriters/PrecedenceLinter.cs [hunk 2]
```diff
@@ -107,27 +111,27 @@ class PrecedenceLinterVisitor : TopDownVisitor<LeftMargin> {
       // that is, we inspect line and column information.
 
       if (expr is BinaryExpr bin && (bin.Op == BinaryExpr.Opcode.Imp || bin.Op == BinaryExpr.Opcode.Exp || bin.Op == BinaryExpr.Opcode.Iff)) {
+        // For
```
Comment (robin-aws) [inline Source/DafnyCore/Rewriters/PrecedenceLinter.cs:114]: Was this suggested by `dotnet format` or something like that?

No objection to the comment move, but personally I find the nested ternary operator much more of a readability issue. :)

File: Source/DafnyStandardLibraries/Makefile [hunk 3]
```diff
@@ -11,8 +11,8 @@ DOO_FILE_ARITHMETIC_TARGET=binaries/DafnyStandardLibraries-arithmetic.doo
 all: check-binary test check-format check-examples
 
 verify:
-	$(DAFNY) verify src/dfyconfig.toml
-	$(DAFNY) verify src/dfyconfig-arithmetic.toml
+	$(DAFNY) verify src/DafnyStdLibs/EnableNonLinearArithmetic/dfyconfig.toml
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/Makefile:14]: Good fix. Should probably delegate to a similar `verify` target on the `TargetSpecific` Makefile too for consistency with the other targets.

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/Collections/Seqs.dfy [hunk 4]
```diff
@@ -73,12 +73,23 @@ module DafnyStdLibs.Collections.Seqs {
...
   }
 
+  /* The concatenation of sequences is associative (four arguments). */
+  lemma LemmaConcatIsAssociative2<T>(a: seq<T>, b: seq<T>, c: seq<T>, d: seq<T>)
+    ensures a + b + c + d == a + (b + c + d)
+  {
+  }
+
+  /* The empty sequence is the neutral element of the concatenation operation. */
+  lemma Neutral(l: seq)
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/Collections/Seqs.dfy:89]: nit: IMHO "Identity" is the more common term. Also consider "RightIdentity", or even "EmptySequenceIsRightIdentity".

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/Collections/Seqs.dfy [hunk 5]
```diff
@@ -390,6 +401,16 @@ module DafnyStdLibs.Collections.Seqs {
   {
   }
 
+  /* If a predicate is true for every member of a sequence as a collection,
+     it is true at every index of the sequence.
+     Useful for converting quantifiers between the two forms
+     to satisfy a precondition in the latter form. */
+  lemma MembershipImpliesIndexing<T>(p: T -> bool, xs: seq<T>)
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/Collections/Seqs.dfy:408]: NICE, amazed we hadn't added this already

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1514.dfy [hunk 6]
```diff
@@ -1,4 +1,5 @@
-// RUN: %testDafnyForEachCompiler --refresh-exit-code=0 "%s" -- --standard-libraries --relax-definite-assignment
+// RUN: %verify "%s" --standard-libraries:true  > "%t"
```
Comment (robin-aws) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1514.dfy:1]: We definitely shouldn't do this as it's weakening the coverage of these tests quite a bit. Are you still hitting errors in the old version even with the patches now in this PR?

**Review by robin-aws — CHANGES_REQUESTED** [review]
Very happy to see this land. Just reviewed the changes outside of the actual library for now, will review the actual content in a second pass later.

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Deserializer.dfy [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+    type Char = byte
+
+    const chars := [
+      '0' as uint8, '1' as uint8, '2' as uint8, '3' as uint8,
+      '4' as uint8, '5' as uint8, '6' as uint8, '7' as uint8,
+      '8' as uint8, '9' as uint8
+    ]
+
+    const charToDigit := map[
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Deserializer.dfy:122]: This probably belongs in the Strings library now instead, but fine to move it there later

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Serializer.dfy [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      Success(Grammar.JString(Grammar.DOUBLEQUOTE, View.OfBytes(bs), Grammar.DOUBLEQUOTE))
+    else
+      Failure(o.error)
+  }
+
+  function Sign(n: int): jminus {
+    View.OfBytes(if n < 0 then ['-' as byte] else [])
+  }
+
+  module ByteStrConversion refines Strings.ParametricConversion {
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Serializer.dfy:48]: Oh but there's also two copies of it? Okay let's move it to Strings before merging unless that's really difficult for some reason. :)

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Spec.dfy [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  import Collections.Seqs
+  import opened BoundedInts
+  import opened Strings
+  import opened Values
+  import opened Wrappers
+  import opened Errors
+  import opened Unicode.UnicodeStringsWithUnicodeChar
+  import opened Arithmetic.Logarithm
+
+  type bytes = seq<uint8>
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Spec.dfy:19]: Shadows the definition from `BoundedInts`, can remove?

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Utils/Views.Writers.dfy [hunk 10]
```diff
@@ -0,0 +1,122 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT 
+ *******************************************************************************/
+
+/**
+ Implements slices over byte strings whose bounds are representable as `int32` native integers.
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Utils/Views.Writers.dfy:7]: Duplicate comment with Views.dfy

**Review by robin-aws — COMMENTED** [review]
Finished reviewing the rest, only spotted a couple of minor additional things (since I reviewed all of this when it first landed in `dafny-lang/libraries` already)

File: Source/DafnyCore/Rewriters/PrecedenceLinter.cs (hunk 2 above)
Comment (stefan-aws) [reply Source/DafnyCore/Rewriters/PrecedenceLinter.cs:114]: No, I had to format it that way to bypass this bug: https://github.com/dotnet/format/issues/1566

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1514.dfy (hunk 6 above)
Comment (stefan-aws) [reply Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1514.dfy:1]: Still hitting errors, see e.g. https://github.com/dafny-lang/dafny/actions/runs/7101157152/job/19328811385?pr=4754. Keeping my fix for now.

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Serializer.dfy (hunk 8 above)
Comment (stefan-aws) [reply Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Serializer.dfy]: Ran into an issue when trying to move it into `Strings`, thus created a JSON submodule `JSON.ByteStrConversion`.

**Review by robin-aws — APPROVED** [review]
Cut https://github.com/dafny-lang/dafny/issues/4847 and worked around it so we could restore the tests. Great work!

**Pull request merged**